### DISCUSSION TITLE
単体テストの考え方/使い方 11章

### DISCUSSION THREAD

**Discussion by y-kaneda** [atman-inc/training#106] [General]
輪読会のサマリ用ディスカッション。
11章分の各章単位程度でコメントを記載するので、不明点、相談点、確認点、勉強になったポイントなどは、各自ディスカッションのスレッドへコメントをお願いします。

**Comment by y-kaneda** [comment]
# 第11章 単体テストのアンチ・パターン

**Comment by y-kaneda** [comment]
## 11.1 プライベートなメソッドに対する単体テスト

**Comment by y-kaneda** [comment]
## 11.1.1 プライベートなメソッドとテストの壊れやすさ

**Comment by y-kaneda** [comment]
### 11.1.2 プライベートなメソッドとテストの不十分な網羅率

**Comment by y-kaneda** [comment]
### 11.1.3 プライベートなメソッドを直接テストすることが受け入れられるのはどのような場合なのか？

**Comment by y-kaneda** [comment]
## 11.2 プライベートな状態の公開

**Comment by y-kaneda** [comment]
## 11.3 テストへのドメイン知識の漏洩

**Reply by nmurakami0** [reply]
例えば何らかのロジックで得たエンティティの "${zip_code}_${address}"みたいな値がactualのときにも、
"2220002_神奈川"みたいにベタ書きしたほうがよい？

**Comment by y-kaneda** [comment]
## 11.4 プロダクション・コードへの汚染

**Reply by nmurakami0** [reply]
設定値で本番環境かのフラグを見ることがあると思うが、それは許容？
(テストかをif文で使うとだめ？)

**Comment by y-kaneda** [comment]
## 11.5 具象クラスに対するテスト・ダブル

**Reply by nmurakami0** [reply]
- クラス自体をスタブに置き換えられないということは、単一責任の原則を守っていないクラスだから

**Comment by y-kaneda** [comment]
## 11.6 単体テストにおける現在日時の扱い

**Reply by nmurakami0** [reply]
- rspecだとtravel_toがあるが、たしかに他の言語だと無くて困ることがある

**Comment by y-kaneda** [comment]
### 11.6.1 環境コンテキスト（ambient context）として現在日時を扱う方法

**Reply by nmurakami0** [reply]
- この実装だと、DateTimeServer.Nowと呼べば、現在日時を得られる？
- アンチパターンである
  - プロダクションコードを汚してしまい、テストの実施が難しくなる
  - staticなフィールドだとテストケース間で共有される依存を持ち込んでしまう

**Comment by y-kaneda** [comment]
### 11.6.2 明示的な依存として現在日時を扱う方法

**Reply by nmurakami0** [reply]
- サービスとして注入するか、値として注入するかの2種類がある
  - 後者の方が扱いやすいのでおすすめ
  - ただ、DIで値オブジェクトを扱えないことがあるので、controllernなどにサービスとして注入して、その後はサービスから得た値を他のクラスやメソッドに渡せば良い
- この例だと、テストではIDateTImeServerを実装したTestDateTimeServerみたいなのを用意して、Nowメソッドでは固定のDateTimeインスタンスを返せば良い

**Comment by y-kaneda** [comment]
## 11.7 結論

**Comment by y-kaneda** [comment]
## まとめ
